thorin/def.cpp

```diff
@@ -73,107 +75,162 @@ UMax::UMax(World& world, Defs ops)
+template<bool up> Ref TExt  <up>::rebuild(World& w, Ref t, Defs  ) const { return w.ext  <up>(t)->set(dbg()); }
+template<bool up> Ref TBound<up>::rebuild(World& w, Ref  , Defs o) const { return w.bound<up>(o)->set(dbg()); }
 
 /*
  * stub
  */
 
-Arr*       Arr   ::stub_(World& w, Ref t) { return w.mut_arr  (t); }
-Global*    Global::stub_(World& w, Ref t) { return w.global(t, is_mutable()); }
-Infer*     Infer ::stub_(World& w, Ref t) { return w.mut_infer(t); }
-Lam*       Lam   ::stub_(World& w, Ref t) { return w.mut_lam  (t->as<Pi>()); }
-Pack*      Pack  ::stub_(World& w, Ref t) { return w.mut_pack (t); }
-Pi*        Pi    ::stub_(World& w, Ref t) { return w.mut_pi   (t, implicit()); }
-Sigma*     Sigma ::stub_(World& w, Ref t) { return w.mut_sigma(t, num_ops()); }
+Arr*    Arr   ::stub(World& w, Ref t) { return w.mut_arr  (t)                ->set(dbg()); }
+Global* Global::stub(World& w, Ref t) { return w.global   (t, is_mutable())  ->set(dbg()); }
+Infer*  Infer ::stub(World& w, Ref t) { return w.mut_infer(t)                ->set(dbg()); }
+Lam*    Lam   ::stub(World& w, Ref t) { return w.mut_lam  (t->as<Pi>())      ->set(dbg()); }
+Pack*   Pack  ::stub(World& w, Ref t) { return w.mut_pack (t)                ->set(dbg()); }
+Pi*     Pi    ::stub(World& w, Ref t) { return w.mut_pi   (t, is_implicit()) ->set(dbg()); }
+Sigma*  Sigma ::stub(World& w, Ref t) { return w.mut_sigma(t, num_ops())     ->set(dbg()); }
 
-template<bool up> TBound<up>* TBound<up>::stub_(World& w, Ref t) { return w.mut_bound<up>(t, num_ops()); }
-template<bool up> TExt  <up>* TExt  <up>::stub_(World&  , Ref  ) { unreachable(); }
+template<bool up> TBound<up>* TBound<up>::stub(World& w, Ref t) { return w.mut_bound<up>(t, num_ops()); }
+template<bool up> TExt  <up>* TExt  <up>::stub(World&  , Ref  ) { unreachable(); }
 
 /*
  * instantiate templates
  */
 
-template Ref TExt  <false>::rebuild_(World&, Ref, Defs) const;
-template Ref TExt  <true >::rebuild_(World&, Ref, Defs) const;
-template Ref TBound<false>::rebuild_(World&, Ref, Defs) const;
-template Ref TBound<true >::rebuild_(World&, Ref, Defs) const;
-template TBound<false>* TBound<false>::stub_(World&, Ref);
-template TBound<true >* TBound<true >::stub_(World&, Ref);
-template TExt  <false>* TExt  <false>::stub_(World&, Ref);
-template TExt  <true >* TExt  <true >::stub_(World&, Ref);
+#ifndef DOXYGEN
+template Ref            TExt<false>  ::rebuild(World&, Ref, Defs) const;
+template Ref            TExt<true >  ::rebuild(World&, Ref, Defs) const;
+template Ref            TBound<false>::rebuild(World&, Ref, Defs) const;
+template Ref            TBound<true >::rebuild(World&, Ref, Defs) const;
+template TBound<false>* TBound<false>::stub(World&, Ref);
+template TBound<true >* TBound<true >::stub(World&, Ref);
+template TExt<false>*   TExt<false>  ::stub(World&, Ref);
+template TExt<true >*   TExt<true >  ::stub(World&, Ref);
+#endif
+
 // clang-format on
 
 /*
- * restructure
+ * freeze
```

Comment:
`freeze` constructs an immutable (formerly structural) from mutable (formerly nominal) by rewrapping its arguments and possibly simplifying it along the way?

Reply:
It *tries* to make an immutable from a mutable. This usually works if the mutable isn't recursive and its var isn't used.
Will add it to the docs.

Reply:
`freeze` was a bit cryptic. It's now called `immutabilize`.